README.md
# react-native-test-runner

[![npm version][npm-image]][npm-url] [![ci][github-ci-image]][github-ci-url]

[npm-url]:https://www.npmjs.com/package/react-native-test-runner
[npm-image]:https://img.shields.io/npm/v/react-native-test-runner.svg
[github-ci-url]:https://github.com/acostalima/react-native-test-runner/actions
[github-ci-image]:https://github.com/acostalima/react-native-test-runner/workflows/Node%20CI/badge.svg

> Run tests in React Native's environment.

## Installation

```sh
$ npm install -D react-native-test-runner
```

## Features 

- Supported test runners:
    - [zora](https://github.com/lorenzofox3/zora)
    - [mocha](https://github.com/mochajs/mocha/)
- Supported assertion libraries:
    - [chai](https://github.com/chaijs/chai)
- React Native version selection.
- iOS runtime and simulator selection.
- Android emulator (AVD) selection.
- Support to use and test libraries with native modules.

## Limitations

- No coverage output.
- Stack traces are not source mapped.
- No support for [React Native Windows](https://github.com/microsoft/react-native-windows) and [React Native macOS](https://github.com/microsoft/react-native-macos).
- JavaScriptCore (JSC) engine only on both Android and iOS.
- No TypeScript (TS) support.

## Usage

```
Usage
    $ rn-test --platform ['ios' | 'android'] [--simulator ios_simulator | --emulator android_avd] [globs ...]

    Default glob: **/test?(s)/**/?(*.)+(spec|test).js.

Options
    --plaform, -p          Platform on which to run the test suite on. One of: 'ios', 'android'.
    --simulator, -s        iOS simulator to run the test suite on.
    --emulator, -e         Android emulator or virtual device (AVD) to run the test suite on.
    --metro-port, -p       Port on which Metro's server should listen to. [Default: 8081]
    --cwd                  Current directory. [Default: process.cwd()]
    --rn                   React Native version to use. [Default: 0.63.4]
    --runner               Test runner to use. One of: 'zora', 'mocha'. [Default: 'zora']
    --require              Path to the module to load before the test suite. If not absolute, cwd is used to resolve the path.
    --app                  Path to the React Native test app root. [Default: ~/.rn-test-app]

Examples
    # Run tests on iPhone 11 simulator with iOS version 14.1 runtime
    $ rn-test --platform ios --simulator 'iPhone 11 (14.1)' 'test/**/*.test.js'

    # Run tests on iPhone 11 simulator with whatever iOS version is available
    $ rn-test --platform ios --simulator 'iPhone 11' 'test/**/*.test.js'

    # Run tests on iOS simulator by UUID
    $ rn-test --platform ios --simulator 'DAC8A157-A03E-4E35-92E6-90752F95BB7A' 'test/**/*.test.js'

    # Run tests on Android emulator
    $ rn-test --platform android --emulator 'Pixel_API_28_AOSP' 'test/**/*.test.js'

Notes
    Do not use shell expansion for globs. Always wrap them in a string.
    $ rn-test 'test/**' ✅
    $ rn-test test/** ❌

    Check out the README for information about advanced options.
```

### Device orchestration

Both `simulator` and `emulator` options are optional.

When they're not passed, the CLI attempts to find the currently booted device. If none or more than one is found, the CLI exits with an error. If only one is found, the CLI installs and launches the test app in that device. The device is not shutdown before the CLI exits. This feature might be useful when running in CI servers which are often responsible for setting up the appropriate test environment.

When they're passed, unless the device cannot be found in the system, the CLI boots the device and shuts it down automatically before exiting. In case of iOS, if the passed simulator is already booted, the CLI behaves exactly as if it was not, i.e., the device is not shutdown before the CLI exists.

### Advanced options

The following options are only available via configuration file loaded by [lilconfig](https://github.com/antonk52/lilconfig).
All CLI options are supported in the configuration file approach as well, but the options passed to the CLI override those defined in the file.  

Example `rn-test.config.js`:

```js
{
    platform: "ios",
    simulator: "iPhone 11 (14.1)",
    runner: "mocha"
}
```

Run the CLI:

```
$ rn-test 'test/**/*.test.js'
```

#### modules

Type: `array`

Install packages in the test app. This is mostly useful when you are using modules that have a native iOS and/or Android component or the code under test depends on them.

Example:

```js
module.exports = {
    modules: ["react-native-get-random-values"]
}
```

#### patches

Type: `array`

Paths to the patch files to apply to the test app.
This might be useful if you need to patch the React Native's source temporarily to fix a bug or add missing functionality.

Example:

```js
module.exports = {
    "patches": [{
        path: require.resolve("react-native-polyfill-globals/patches/react-native+0.63.3.patch")
    }]
}
```

If not absolute, the current working directory (`cwd`) is used to resolve the path. Alternatively, you can specify your own `cwd` for the each patch to apply:

```js
module.exports = {
    "patches": [{
        path: "node_modules/react-native-polyfill-globals/patches/react-native+0.63.3.patch"
        cwd: process.cwd()
    }]
}
```

## Tests

This section is about how you can run the test suite locally.

### Environment setup

The test suite can be parameterized with the following environment variables:

| Variable | CLI option | Default |
|----------|------------|---------|
| `REACT_NATIVE_VERSION` | `rn` | CLI default |
| `REACT_NATIVE_TEST_APP` | `app` | CLI default |
| `IOS_SIMULATOR` | `simulator` | - |
| `ANDROID_SIMULATOR`| `emulator` | - |

Both `IOS_SIMULATOR` and `ANDROID_EMULATOR` must be set according to what iOS simulators and Android emulators you have installed in your system.

You can find the list of iOS simulators available in your system by running the following command:

```
$ xcrun xctrace list devices
```

For the list of Android emulators, run:

```
$ emulator -list-avds
```

### Run on iOS

⚠️ A machine with macOS operating system is required!

```
$ npm t -- ios
```

### Run on Android

```
$ npm t -- android
```

## Known issues

- While `metroPort` works in [Android](https://github.com/facebook/react-native/pull/23616), it doesn't on [iOS](https://github.com/facebook/react-native/issues/9145). Metro does listen to the specified port but, the app, as a client, still attempts to load the bundle from port 8081.
## License

MIT © [André Costa Lima](https://github.com/acostalima)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [5.0.0](https://github.com/acostalima/react-native-test-runner/compare/v4.0.0...v5.0.0) (2021-03-30)


### ⚠ BREAKING CHANGES

* improve device management

### Features

* add option to reset metro's cache ([cdf241d](https://github.com/acostalima/react-native-test-runner/commit/cdf241da52a73d71782cbafb7d10b0895ddae463))


### Bug Fixes

* improve device management ([7585d03](https://github.com/acostalima/react-native-test-runner/commit/7585d03ed7cc9c2a8bc4be0ad185199e42f398d8))

## [4.0.0](https://github.com/acostalima/react-native-test-runner/compare/v3.0.2...v4.0.0) (2021-03-25)


### ⚠ BREAKING CHANGES

* renamed configFile option to config

* fix ios workflow ([2e2045b](https://github.com/acostalima/react-native-test-runner/commit/2e2045b7d1c02ccca8bb1fd73067def0d86b3f2a))

### [3.0.2](https://github.com/acostalima/react-native-test-runner/compare/v3.0.1...v3.0.2) (2021-03-03)

### [3.0.1](https://github.com/acostalima/react-native-test-runner/compare/v3.0.0...v3.0.1) (2021-03-03)


### Bug Fixes

* pick up test files located at cwd ([6c7a1f5](https://github.com/acostalima/react-native-test-runner/commit/6c7a1f5d1e078faec4179a79f5f33e7a53a0e140))

## [3.0.0](https://github.com/acostalima/react-native-test-runner/compare/v2.3.0...v3.0.0) (2021-03-02)


### ⚠ BREAKING CHANGES

* changed nativeModules option to just modules

### Bug Fixes

* modules installed in the test app are now resolved by metro ([b984fa9](https://github.com/acostalima/react-native-test-runner/commit/b984fa95446dd1612f7cc2fbd1f16a962f4137ea))

## [2.3.0](https://github.com/acostalima/react-native-test-runner/compare/v2.2.1...v2.3.0) (2021-01-31)


### Features

* add support to load and use process.env in the test app ([642d9b7](https://github.com/acostalima/react-native-test-runner/commit/642d9b70f030fe32909842f2e46328ad843a2035))

### [2.2.1](https://github.com/acostalima/react-native-test-runner/compare/v2.2.0...v2.2.1) (2021-01-28)


### Bug Fixes

* handle patch fail error correctly ([eddbccf](https://github.com/acostalima/react-native-test-runner/commit/eddbccffb5389d26e8ee78ee3bb8ee9dea8e420f))

## [2.2.0](https://github.com/acostalima/react-native-test-runner/compare/v2.1.0...v2.2.0) (2021-01-28)


### Features

* proceed on patch apply fail and allow to set cwd for patch ([e1f428c](https://github.com/acostalima/react-native-test-runner/commit/e1f428ce4166ca74b0669cb5a742d69d7edd1379))

## [2.1.0](https://github.com/acostalima/react-native-test-runner/compare/v2.0.0...v2.1.0) (2021-01-28)


### Features

* find project root in mono repositories ([495ff4a](https://github.com/acostalima/react-native-test-runner/commit/495ff4aa432124de3422d077d001022fc1daabd6))
* preload module, native modules and patches support ([3c66faf](https://github.com/acostalima/react-native-test-runner/commit/3c66faf9825551318ea2970d94f578c32631d518))

## [2.0.0](https://github.com/acostalima/react-native-test-runner/compare/v1.0.10...v2.0.0) (2021-01-25)


### ⚠ BREAKING CHANGES

* zora and mocha are dev dependencies

### Features

* add support for mocha test runner ([c9424ba](https://github.com/acostalima/react-native-test-runner/commit/c9424bab29c238621aedceb73cbc9ca5d0a3336f))

### [1.0.10](https://github.com/acostalima/react-native-test-runner/compare/v1.0.9...v1.0.10) (2021-01-11)


### Bug Fixes

* reset metro cache only on ci ([45cc6b0](https://github.com/acostalima/react-native-test-runner/commit/45cc6b09323beef9a8372fa8be5c1d357bfb9b1c))
* resolve module request for ./index should only target app root ([f0354c2](https://github.com/acostalima/react-native-test-runner/commit/f0354c2bd633c8474c1e4d7d020b3d459d8325ea))
* test results logging should receive metro client_log event data ([0ee436a](https://github.com/acostalima/react-native-test-runner/commit/0ee436a82d2de71548310470b7679c8a57764336))

### [1.0.9](https://github.com/acostalima/react-native-test-runner/compare/v1.0.8...v1.0.9) (2021-01-11)


### Bug Fixes

* do not log empty error data ([061a446](https://github.com/acostalima/react-native-test-runner/commit/061a446ad8ad76d96ebf97f92a7332001a3fd2c1))
* fail and pass check given output message ([b7898ae](https://github.com/acostalima/react-native-test-runner/commit/b7898ae99577598c5872dd08fd7226e789bb8c62))
* log errors captured by zora ([10a65cc](https://github.com/acostalima/react-native-test-runner/commit/10a65cc0e7cc1d4d99f72bdb1cb1c2e79d5fd9dd))
* resolve module requests for react native modules in metro ([abe3b64](https://github.com/acostalima/react-native-test-runner/commit/abe3b648529a35c0878e7f7b1963f9146e24b1e5))

### [1.0.8](https://github.com/acostalima/react-native-test-runner/compare/v1.0.7...v1.0.8) (2021-01-07)


### Bug Fixes

* add cwd to metro watch folders ([aa14795](https://github.com/acostalima/react-native-test-runner/commit/aa147959fca6c965d719a7372dcd1e23d079596d))

### [1.0.7](https://github.com/acostalima/react-native-test-runner/compare/v1.0.6...v1.0.7) (2021-01-07)


### Bug Fixes

* add zora to metro module resolution ([aefe141](https://github.com/acostalima/react-native-test-runner/commit/aefe14177d93bae5121bc2654db09793ca545d2e))

### [1.0.6](https://github.com/acostalima/react-native-test-runner/compare/v1.0.5...v1.0.6) (2021-01-07)


### Bug Fixes

* add babel dependencies to metro module resolution ([57d9c48](https://github.com/acostalima/react-native-test-runner/commit/57d9c48447441a37dcee2ae290335941f7e943b0))

### [1.0.5](https://github.com/acostalima/react-native-test-runner/compare/v1.0.4...v1.0.5) (2021-01-07)


### Bug Fixes

* test app is now created by the cli and is no longer packaged ([f85f071](https://github.com/acostalima/react-native-test-runner/commit/f85f07161646664ed6e83aa34c87f1bac05f77fd))


cli/index.js
#!/usr/bin/env node

'use strict';

const path = require('path');
const os = require('os');
const meow = require('meow');
const semver = require('semver');
const { lilconfigSync: loadConfig } = require('lilconfig');
const mergeOptions = require('merge-options');
const runIOS = require('./run-ios');
const runAndroid = require('./run-android');
const runMetroServer = require('./metro-server');
const createTestApp = require('./create-test-app');
const createTestRunner = require('./test-runners');
const { findMonoRepoRoot } = require('./utils');

const cli = meow(`
Usage
    $ rn-test --platform ['ios' | 'android'] [--simulator ios_simulator | --emulator android_avd] [globs ...]

    Default glob: **/test?(s)/**/?(*.)+(spec|test).js.

Options
    --plaform, -p          Platform on which to run the test suite on. One of: 'ios', 'android'.
    --simulator, -s        iOS simulator to run the test suite on.
    --emulator, -e         Android emulator or virtual device (AVD) to run the test suite on.
    --metro-port, -p       Port on which Metro's server should listen to. [Default: 8081]
    --cwd                  Current directory. [Default: process.cwd()]
    --rn                   React Native version to use. [Default: 0.63.4]
    --runner               Test runner to use. One of: 'zora', 'mocha'. [Default: 'zora']
    --require              Path to the module to load before the test suite. If not absolute, cwd is used to resolve the path.
    --app                  Path to the React Native test app root. [Default: ~/.rn-test-app]

Examples
    # Run tests on iPhone 11 simulator with iOS version 14.1 runtime
    $ rn-test --platform ios --simulator 'iPhone 11 (14.1)' 'test/**/*.test.js'

    # Run tests on iPhone 11 simulator with whatever iOS version is available
    $ rn-test --platform ios --simulator 'iPhone 11' 'test/**/*.test.js'

    # Run tests on iOS simulator by UUID
    $ rn-test --platform ios --simulator 'DAC8A157-A03E-4E35-92E6-90752F95BB7A' 'test/**/*.test.js'

    # Run tests on Android emulator
    $ rn-test --platform android --emulator 'Pixel_API_28_AOSP' 'test/**/*.test.js'

Notes
    Do not use shell expansion for globs. Always wrap them in a string.
    $ rn-test 'test/**' ✅
    $ rn-test test/** ❌

    Check out the README for information about advanced options.
`,
{
    flags: {
        platform: {
            type: 'string',
            alias: 'p',
        },
        simulator: {
            type: 'string',
            alias: 's',
        },
        emulator: {
            type: 'string',
            alias: 'e',
        },
        metroPort: {
            type: 'number',
            alias: 'p',
        },
        cwd: {
            type: 'string',
        },
        rn: {
            type: 'string',
        },
        runner: {
            type: 'string',
        },
        require: {
            type: 'string',
            alias: 'r',
        },
        config: {
            type: 'string',
            alias: 'c',
        },
        app: {
            type: 'string',
        },
    },
});

const CLI_DEFAULT_OPTIONS = {
    metroPort: 8081,
    cwd: process.cwd(),
    rn: '0.63.4',
    runner: 'zora',
    app: path.join(os.homedir(), '.rn-test-app'),
};
const SUPPORTED_PLATFORMS = ['android', 'ios'];
const SUPPORTED_RUNNERS = ['zora', 'mocha'];

const fileConfigExplorer = loadConfig('rn-test', { stopDir: process.cwd() });
const fileConfigSearchResult = fileConfigExplorer.search();
const fileConfig = cli.flags.config ? fileConfigExplorer.load(cli.flags.config) : fileConfigSearchResult || {};
const options = mergeOptions(CLI_DEFAULT_OPTIONS, fileConfig.config, cli.flags);

if (!SUPPORTED_RUNNERS.includes(options.runner)) {
    console.error(`Unknown runner: ${options.runner}. Supported runners are: ${SUPPORTED_RUNNERS.join(', ')}.`);
    process.exit(2);
}

if (!SUPPORTED_PLATFORMS.includes(options.platform)) {
    console.error(`Unknown platform: ${options.platform}. Supported platforms are: ${SUPPORTED_PLATFORMS.join(', ')}.`);
    process.exit(2);
}

if (!semver.valid(options.rn)) {
    console.error(`Invalid React Native version scheme: ${options.rn}. Must be compliant with Semantic Version.`);
    process.exit(2);
}

const runTestApp = (options) => {
    options = {
        ...options,
        testAppRoot: path.join(options.app, options.platform),
    };

    switch (options.platform) {
    case 'android':
        return runAndroid(options);
    case 'ios':
        return runIOS(options);
    default:
        throw new Error(`Developer error. Unknown platform: ${options.platform}`);
    }
};

const runTests = async (options, testFileGlobs) => {
    try {
        const monoRepoRoot = await findMonoRepoRoot(options.cwd);
        const jsAppRoot = path.join(__dirname, '..');
        const testRunner = createTestRunner(options, testFileGlobs);

        await createTestApp(options);

        await runMetroServer({
            cwd: options.cwd,
            port: options.metroPort,
            testAppRoot: options.app,
            monoRepoRoot,
            jsAppRoot,
            testRunner,
            testAppUserModules: options.modules,
        });

        const shutdownNativePlatform = await runTestApp(options);

        await testRunner.reporter.waitForTests();
        await shutdownNativePlatform();

        process.exit(testRunner.reporter.didPass() ? 0 : 1);
    } catch (error) {
        console.error(error);
        process.exit(3);
    }
};

runTests(options, cli.input);
